apps/op-app/src/use-cases/audit.ts

```diff
@@ -0,0 +1,100 @@
+import {
+  Session,
+  SessionEnvironment,
+  eventsSchema,
+  getEvent,
+  getSession,
+  writeEvent,
+} from "@/domain/session.js";
+import { StorageEnvironment, sendEventsMessage } from "@/domain/storage.js";
+import * as O from "fp-ts/lib/Option.js";
+import * as RTE from "fp-ts/lib/ReaderTaskEither.js";
+import { flow, pipe } from "fp-ts/lib/function.js";
+import { RPParams, auditEventSchema } from "io-fims-common/domain/audit-event";
+import * as jose from "jose";
+import * as assert from "node:assert/strict";
+
+export class AuditError extends Error {
+  name = "AuditError";
+  constructor(message?: string) {
+    super(message || "Unexpected error during audit use case.");
+  }
+}
+
+type Context = SessionEnvironment & StorageEnvironment;
+
+const safeFindSession = flow(
+  getSession,
+  RTE.map(
+    flow(
+      O.map(({ userMetadata }: Session) => userMetadata),
+      O.toUndefined,
+    ),
+  ),
+);
+
+const safeGetEvent = flow(getEvent, RTE.map(O.toUndefined));
+
+export class AuditUseCase {
```

Comment:
> A use case is a piece of business logic that represents a single task that the system needs to perform. The use case encapsulates the rules and logic required to perform the task, and defines the inputs and outputs required for the operation.

https://nanosoft.co.za/blog/post/clean-architecture-use-cases

----

so they are not "services" classes that group related function, they must represent only one action